wire/internal/wire/wire.go

```diff
@@ -63,34 +64,37 @@ func (gen GeneratedFile) Commit() error {
 // env is nil or empty, it is interpreted as an empty set of variables.
 // In case of duplicate environment variables, the last one in the list
 // takes precedence.
-func Generate(ctx context.Context, wd string, env []string, pkgPattern string) (GeneratedFile, []error) {
-	pkgs, errs := load(ctx, wd, env, []string{pkgPattern})
+func Generate(ctx context.Context, wd string, env []string, patterns []string) ([]GeneratedFile, []error) {
+	pkgs, errs := load(ctx, wd, env, patterns)
 	if len(errs) > 0 {
-		return GeneratedFile{}, errs
+		return nil, errs
 	}
-	if len(pkgs) != 1 {
-		// This is more of a violated precondition than anything else.
-		return GeneratedFile{}, []error{fmt.Errorf("load: got %d packages", len(pkgs))}
-	}
-	outDir, err := detectOutputDir(pkgs[0].GoFiles)
-	if err != nil {
-		return GeneratedFile{}, []error{fmt.Errorf("load: %v", err)}
-	}
-	outFname := filepath.Join(outDir, "wire_gen.go")
-	g := newGen(pkgs[0])
-	injectorFiles, errs := generateInjectors(g, pkgs[0])
-	if len(errs) > 0 {
-		return GeneratedFile{}, errs
-	}
-	copyNonInjectorDecls(g, injectorFiles, pkgs[0].TypesInfo)
-	goSrc := g.frame()
-	fmtSrc, err := format.Source(goSrc)
-	if err != nil {
-		// This is likely a bug from a poorly generated source file.
-		// Return an error and the unformatted source.
-		return GeneratedFile{Path: outFname, Content: goSrc}, []error{err}
+	generated := make([]GeneratedFile, 0, len(pkgs))
+	ec := new(errorCollector)
```

Comment:
Alternatively, we could always return exactly one `GeneratedFile` per package, remove the `errors` return value from this function, and add an `Err` field to `GeneratedFile`. That would make it easier to identify which package each error was for. Agree?

Reply:
I mostly agree, but I think we would need to shift to returning package-centric things. I don't think we can entire drop the separate error return, because errors in unrelated packages or the load process could cause a failure that's not logically related to the other packages.

Reply:
OK, how about changing `Generate` to return a slice of `GenerateResult` plus an error? `GenerateResult` is just `GeneratedFile` renamed, plus a slice of errors.

Reply:
SGTM

Reply:
PTAL.